scripts/speech_recognition/estimate_data_weights.py

```diff
@@ -0,0 +1,172 @@
+        for subentry in entry:
+            validate(subentry, _level + 1)
+        return
+
+    assert "type" in entry, f"Invalid YAML data config at nesting level {_level}: missing key 'type' in entry={entry}"
+
+    if entry.type == "group":
+        for subentry in entry["input_cfg"]:
+            validate(subentry, _level + 1)
+
+
+def count(entry: DictConfig | ListConfig, weight_key: str) -> None:
+    if isinstance(entry, ListConfig):
+        for subentry in entry:
+            count(subentry, weight_key=weight_key)
+        return
+    if entry.type == "group":
+        for subentry in entry["input_cfg"]:
+            count(subentry, weight_key=weight_key)
+        return
+
+    with quick_iter_options(entry):
+        iterable, is_tarred = get_parser_fn(entry.type)(entry)
+        stats = {"num_hours": 0.0, "num_examples": 0}
+        for example in iterable:
+            if hasattr(example, "duration"):
+                stats["num_hours"] += example.duration
+            stats["num_examples"] += 1
+        stats["num_hours"] /= 3600.0
+
+    if weight_key == "num_hours" and stats[weight_key] == 0.0:
+        raise RuntimeError(
+            f"Cannot set weights based on 'num_hours': at least one dataset has examples without 'duration' property. "
+            f"Details: {entry=}"
+        )
+
+    entry["weight"] = stats[weight_key]
+
+
+def aggregate_group_weights(entry: DictConfig | ListConfig) -> None:
+    if isinstance(entry, ListConfig):
+        for subentry in entry:
+            aggregate_group_weights(subentry)
+        return
+
+    if entry.type != "group":
+        return
+
+    for subentry in entry["input_cfg"]:
+        if "weight" not in subentry:
+            aggregate_group_weights(subentry)
+
+    entry.weight = sum(subentry["weight"] for subentry in entry["input_cfg"])
+
+
+def reweight(entry: DictConfig | ListConfig, temperature: None | float | list[float]) -> None:
+    if not temperature or (isinstance(entry, DictConfig) and entry.type != "group"):
+        return
+
+    if isinstance(temperature, Sequence):
+        temperature, *next_temperatures = temperature
+    else:
+        next_temperatures = temperature
+
+    if isinstance(entry, ListConfig):
+        for subentry in entry:
+            reweight(subentry, temperature=next_temperatures)
+        new_weights = temperature_reweighting([se.weight for se in entry], temperature=temperature)
+        for se, nw in zip(entry, new_weights):
+            se.weight = nw
+        return
+
+    for subentry in entry["input_cfg"]:
+        reweight(subentry, temperature=next_temperatures)
+
+    new_weights = temperature_reweighting([se.weight for se in entry["input_cfg"]], temperature=temperature)
+    for se, nw in zip(entry["input_cfg"], new_weights):
+        se.weight = nw
+
+
+def temperature_reweighting(weights: list[float], temperature: float = 1.0):
+    """(w_i ^ alpha / sum(w_i ^ alpha))"""
+    weights = np.asarray(weights) ** temperature
```

Comment:
Would you mind adding an option for `log` scaling of weights to reduce sensitivity? (Some grouping will have weight difference by orders of magnitude so base exponent scaling may be too volatile as we scale.)